[WPE] Implement ImageAdapter

dpino opened the pull request:
#### 488755eb5494242e491ff5220ce560c0b4321edb
<pre>
[WPE] Implement ImageAdapter
<a href="https://bugs.webkit.org/show_bug.cgi?id=208869">https://bugs.webkit.org/show_bug.cgi?id=208869</a>

Reviewed by Carlos Garcia Campos.

Until now WPE didn&apos;t have a specific implementation of ImageAdapter,
and used a generic one instead.  This affected the results of some
layout tests.

It turns out that some parts of ImageAdapterGtk can be actually
reused for WPE, particularly the code that loads bundled images from
a gresource file.

Also, modify PlatformWPE.cmake to bundle images into gresource file.

* LayoutTests/platform/glib/editing/execCommand/insertImage-expected.txt: Renamed from LayoutTests/platform/gtk/editing/execCommand/insertImage-expected.txt.
* LayoutTests/platform/glib/fast/borders/rtl-border-05-expected.txt: Renamed from LayoutTests/platform/gtk/fast/borders/rtl-border-05-expected.txt.
* LayoutTests/platform/glib/fast/dom/34176-expected.txt: Renamed from LayoutTests/platform/gtk/fast/dom/34176-expected.txt.
* LayoutTests/platform/glib/fast/dom/inner-text-expected.txt: Renamed from LayoutTests/platform/gtk/fast/dom/inner-text-expected.txt.
* LayoutTests/platform/glib/fast/encoding/utf-16-big-endian-expected.txt: Renamed from LayoutTests/platform/gtk/fast/encoding/utf-16-big-endian-expected.txt.
* LayoutTests/platform/glib/fast/encoding/utf-16-little-endian-expected.txt: Renamed from LayoutTests/platform/gtk/fast/encoding/utf-16-little-endian-expected.txt.
* LayoutTests/platform/glib/fast/hidpi/broken-image-icon-hidpi-expected.txt: Renamed from LayoutTests/platform/gtk/fast/hidpi/broken-image-icon-hidpi-expected.txt.
* LayoutTests/platform/glib/fast/hidpi/broken-image-icon-very-hidpi-expected.txt: Renamed from LayoutTests/platform/gtk/fast/hidpi/broken-image-icon-very-hidpi-expected.txt.
* LayoutTests/platform/glib/tables/mozilla/bugs/bug2962-expected.txt: Renamed from LayoutTests/platform/gtk/tables/mozilla/bugs/bug2962-expected.txt.
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/platform/wpe/editing/execCommand/insertImage-expected.txt:
* LayoutTests/platform/wpe/fast/borders/rtl-border-05-expected.txt:
* LayoutTests/platform/wpe/fast/dom/34176-expected.txt:
* LayoutTests/platform/wpe/fast/dom/inner-text-expected.txt:
* LayoutTests/platform/wpe/fast/encoding/utf-16-big-endian-expected.txt:
* LayoutTests/platform/wpe/fast/encoding/utf-16-little-endian-expected.txt:
* LayoutTests/platform/wpe/fast/hidpi/broken-image-icon-hidpi-expected.txt:
* LayoutTests/platform/wpe/fast/hidpi/broken-image-icon-very-hidpi-expected.txt:
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug2962-expected.txt:
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug56201-expected.txt: Removed.
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/ImageAdapter.cpp:
* Source/WebCore/platform/graphics/glib/ImageAdapterGLib.cpp: Copied from Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp.
(WebCore::loadImageFromGResource):
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::invalidate):
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp:
(WebCore::loadImageFromGResource): Deleted.
(WebCore::ImageAdapter::loadPlatformResource): Deleted.
(WebCore::ImageAdapter::invalidate): Deleted.
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Resources/wpe/wpe-theme.css: Added.
(webkitwebview &gt; dimming):
(webkitwebview &gt; shadow):
(webkitwebview &gt; shadow.right):
(webkitwebview &gt; shadow.left):
(webkitwebview &gt; border):
(webkitwebview &gt; outline):

Canonical link: <a href="https://commits.webkit.org/299620@main">https://commits.webkit.org/299620@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36f672dbd0269f57ba103561f265b414b5e5f090

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119630 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39324 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29976 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/125913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71704 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/121506 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/40020 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47902 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/125913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/60160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122581 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/31940 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107263 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/125913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30976 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25367 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69559 "Failed to checkout and rebase branch from PR 50172") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/101406 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25560 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/128878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46552 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/35263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/128878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/46917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103456 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/128878 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22747 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/43116 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19026 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46414 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52120 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/45880 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/49229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/47566 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/97c8de59715a1ac9ca9a02b07f016763641652e9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50172

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/35417fe8bc8034a2bf080f81aab3e9cda2ba38aa. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50172

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a3b0b461cdfe17d7557381e4c7818def8d33b543. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50172

carlosgcampos commented on Source/WebCore/platform/graphics/wpe/ImageAdapterWPE.cpp:
```diff
@@ -0,0 +1,57 @@
+static Ref<Image> loadImageFromGResource(const char* iconName)
+{
+    auto icon = BitmapImage::create();
+    GUniquePtr<char> path(g_strdup_printf(IMAGES_GRESOURCE_PATH "/%s", iconName));
```
I think it's ok to use the path directly here instead of defining a macro, since it's only used here. This code is duplicated in gtk implementation, though, so maybe we can add ImageAdapterGLib.cpp to define loadPlatformResource.

carlosgcampos commented on Source/WebKit/PlatformWPE.cmake:
```diff
@@ -356,8 +356,14 @@ add_custom_command(
 
-set(WebKitResources
-)
+set(WebKitResources "")
+list(APPEND WebKitResources "<file alias=\"css/wpe-theme.css\">wpe-theme.css</file>\n")
+list(APPEND WebKitResources "<file alias=\"images/missingImage\">missingImage.png</file>\n")
+list(APPEND WebKitResources "<file alias=\"images/missingImage@2x\">missingImage@2x.png</file>\n")
+list(APPEND WebKitResources "<file alias=\"images/missingImage@3x\">missingImage@3x.png</file>\n")
+list(APPEND WebKitResources "<file alias=\"images/panIcon\">panIcon.png</file>\n")
+list(APPEND WebKitResources "<file alias=\"images/textAreaResizeCorner\">textAreaResizeCorner.png</file>\n")
+list(APPEND WebKitResources "<file alias=\"images/textAreaResizeCorner@2x\">textAreaResizeCorner@2x.png</file>\n")
```
Why don't you just set in a single command instead?

dpino replied:
I used `PlatformGTK.cmake` as reference. I don't know the reason why in `PlatformGTK.cmake` was done this way, but I suspect it had to do with a lint error returned by `Tools/Scripts/check-webkit-style`. I managed to solve this lint error using a single list, so I changed `PlatformGTK.cmake` too.

carlosgcampos reviewed (commented):
I'm ok with this patch, my only concern is that it increases the binary size, so I would confirm with @magomez before landing this.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/37a139800a9c1362ce4449a323a9f82c23524979. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50172

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ca3beb3044b583db75969b610f79b7780c66e39c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50172

carlosgcampos commented on Source/WebCore/SourcesGTK.txt:
```diff
@@ -79,11 +79,12 @@ platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.cpp @no-unify
 platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
 platform/graphics/gbm/PlatformDisplayGBM.cpp @no-unify
 
+platform/graphics/glib/ImageAdapterGLib.cpp
+platform/graphics/gtk/ImageAdapterGtk.cpp
```
```suggestion
platform/graphics/glib/ImageAdapterGLib.cpp

```

carlosgcampos commented on Source/WebCore/SourcesGTK.txt:
```diff
@@ -79,11 +79,12 @@ platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.cpp @no-unify
 platform/graphics/gtk/GdkCairoUtilities.cpp
 platform/graphics/gtk/GdkSkiaUtilities.cpp
 platform/graphics/gtk/IconGtk.cpp
-platform/graphics/gtk/ImageAdapterGtk.cpp
```
Keep this here, we normally have sources alphabetically sorted

carlosgcampos commented on Source/WebCore/SourcesWPE.txt:
```diff
@@ -65,6 +65,7 @@ platform/graphics/PlatformDisplay.cpp @no-unify
 
 platform/graphics/angle/PlatformDisplayANGLE.cpp @no-unify
 
+platform/graphics/glib/ImageAdapterGLib.cpp
```
```suggestion
platform/graphics/glib/ImageAdapterGLib.cpp

```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/36f672dbd0269f57ba103561f265b414b5e5f090. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50172

dpino committed:
[WPE] Implement ImageAdapter
https://bugs.webkit.org/show_bug.cgi?id=208869

Reviewed by Carlos Garcia Campos.

Until now WPE didn't have a specific implementation of ImageAdapter,
and used a generic one instead.  This affected the results of some
layout tests.

It turns out that some parts of ImageAdapterGtk can be actually
reused for WPE, particularly the code that loads bundled images from
a gresource file.

Also, modify PlatformWPE.cmake to bundle images into gresource file.

* LayoutTests/platform/glib/editing/execCommand/insertImage-expected.txt: Renamed from LayoutTests/platform/gtk/editing/execCommand/insertImage-expected.txt.
* LayoutTests/platform/glib/fast/borders/rtl-border-05-expected.txt: Renamed from LayoutTests/platform/gtk/fast/borders/rtl-border-05-expected.txt.
* LayoutTests/platform/glib/fast/dom/34176-expected.txt: Renamed from LayoutTests/platform/gtk/fast/dom/34176-expected.txt.
* LayoutTests/platform/glib/fast/dom/inner-text-expected.txt: Renamed from LayoutTests/platform/gtk/fast/dom/inner-text-expected.txt.
* LayoutTests/platform/glib/fast/encoding/utf-16-big-endian-expected.txt: Renamed from LayoutTests/platform/gtk/fast/encoding/utf-16-big-endian-expected.txt.
* LayoutTests/platform/glib/fast/encoding/utf-16-little-endian-expected.txt: Renamed from LayoutTests/platform/gtk/fast/encoding/utf-16-little-endian-expected.txt.
* LayoutTests/platform/glib/fast/hidpi/broken-image-icon-hidpi-expected.txt: Renamed from LayoutTests/platform/gtk/fast/hidpi/broken-image-icon-hidpi-expected.txt.
* LayoutTests/platform/glib/fast/hidpi/broken-image-icon-very-hidpi-expected.txt: Renamed from LayoutTests/platform/gtk/fast/hidpi/broken-image-icon-very-hidpi-expected.txt.
* LayoutTests/platform/glib/tables/mozilla/bugs/bug2962-expected.txt: Renamed from LayoutTests/platform/gtk/tables/mozilla/bugs/bug2962-expected.txt.
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/platform/wpe/editing/execCommand/insertImage-expected.txt:
* LayoutTests/platform/wpe/fast/borders/rtl-border-05-expected.txt:
* LayoutTests/platform/wpe/fast/dom/34176-expected.txt:
* LayoutTests/platform/wpe/fast/dom/inner-text-expected.txt:
* LayoutTests/platform/wpe/fast/encoding/utf-16-big-endian-expected.txt:
* LayoutTests/platform/wpe/fast/encoding/utf-16-little-endian-expected.txt:
* LayoutTests/platform/wpe/fast/hidpi/broken-image-icon-hidpi-expected.txt:
* LayoutTests/platform/wpe/fast/hidpi/broken-image-icon-very-hidpi-expected.txt:
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug2962-expected.txt:
* LayoutTests/platform/wpe/tables/mozilla/bugs/bug56201-expected.txt: Removed.
* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/SourcesWPE.txt:
* Source/WebCore/platform/graphics/ImageAdapter.cpp:
* Source/WebCore/platform/graphics/glib/ImageAdapterGLib.cpp: Copied from Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp.
(WebCore::loadImageFromGResource):
(WebCore::ImageAdapter::loadPlatformResource):
(WebCore::ImageAdapter::invalidate):
* Source/WebCore/platform/graphics/gtk/ImageAdapterGtk.cpp:
(WebCore::loadImageFromGResource): Deleted.
(WebCore::ImageAdapter::loadPlatformResource): Deleted.
(WebCore::ImageAdapter::invalidate): Deleted.
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Resources/wpe/wpe-theme.css: Added.
(webkitwebview > dimming):
(webkitwebview > shadow):
(webkitwebview > shadow.right):
(webkitwebview > shadow.left):
(webkitwebview > border):
(webkitwebview > outline):

Canonical link: https://commits.webkit.org/299620@main
